src/actions/attack.cpp

```diff
@@ -1604,6 +1604,63 @@ int under_leadership(const unit &u, const map_location& loc, const_attack_ptr we
 	return leader_effect.get_composite_value();
 }
 
+int attack_type::specials_alignment(int combat_modifier, bool is_fearless) const
+{
+	if(!has_special_or_ability("alignment")){
+		return combat_modifier;
+	}
+	// if "alignment" attribute not eqauls to string in checking_alignment chain,
+	// or not "special_alignment special/abilities then alignment_type returned by default.
+	const std::set<std::string> checking_alignment = {"neutral", "lawful", "chaotic", "liminal"};
+	unit_ability_list abil_list = get_specials_and_abilities("alignment");
+	for(unit_ability_list::iterator i = abil_list.begin(); i != abil_list.end();) {
+		if(checking_alignment.count((*i->ability_cfg)["alignment"].str()) == 0){
+			i = abil_list.erase(i);
+		} else {
+			++i;
+		}
+	}
+	if(abil_list.empty()){
+		return combat_modifier;
+	}
+
+	//if list no empty then lawful_bonus and max_liminal_bonus calculated and aplied to alignment specified
+	assert(display::get_singleton());
```

Comment:
I think the calculations from lines 1629 through 1655 can be put into a separate function `get_lawful_and_max_liminal_bonus(unit_ability_list abil_list)` to make the code more modular.